runtime/src/lib.rs

```diff
@@ -506,6 +527,277 @@ pub trait Blob: Clone + Send + Sync + 'static {
     fn sync(&self) -> impl Future<Output = Result<(), Error>> + Send;
 }
 
+/// Owns a runtime context, allowing it to be temporarily taken and later returned.
+///
+/// This behaves similarly to `Option<C>` but supports implementing external traits
+/// (such as [`RngCore`]) on the wrapper type.
+#[derive(Clone, Debug)]
+pub enum ContextSlot<C> {
```

Comment:
The primary insight of this PR is that we can use an `enum` to represent `Option<Context>` and it can satisfy all interfaces (with no real casting).